File: frontend/src/api/mcpRuntime.ts
// ============================================================
// MCP Runtime API client — read-only runtime surface (Types + API).
//
// Mirrors the backend `/api/mcp/servers` read-only routes. There is
// intentionally NO tool-execution API here: MCP tools execute only through
// the Security Execution Gateway, never from the frontend.
// ============================================================

import { request } from "./client";

// ── Types (match backend serde output) ──

export type McpRuntimeStatus =
  | "disconnected"
  | "connecting"
  | "ready"
  | "degraded"
  | "unavailable"
  | "misconfigured"
  | "disabled";

export type McpProtocolVersion = "2025-11-25" | "2026-07-28";

export interface McpServerCapabilities {
  tools: boolean;
  resources: boolean;
  prompts: boolean;
  listChanged?: boolean;
  extensions?: unknown;
}

export interface McpServerRuntimeSummary {
  id: string;
  name: string;
  transport: string;
  protocol_version: McpProtocolVersion;
  status: McpRuntimeStatus;
  capabilities: McpServerCapabilities;
  tools_count: number;
  resources_count: number;
  resource_templates_count: number;
  prompts_count: number;
  last_refresh?: number | null;
  safe_error?: string | null;
}

export interface McpRuntimeTool {
  name: string;
  title?: string | null;
  description?: string | null;
  inputSchema?: Record<string, unknown> | null;
}

export interface McpResourceDescriptor {
  uri: string;
  name: string;
  title?: string | null;
  description?: string | null;
  mimeType?: string | null;
  size?: number | null;
  annotations?: unknown;
}

export interface McpResourceTemplate {
  uriTemplate: string;
  name: string;
  title?: string | null;
  description?: string | null;
  mimeType?: string | null;
}

/** Wire shape for a single resource content item (tagged `kind`). */
export type McpResourceContent =
  | { kind: "text"; uri: string; mime_type?: string | null; text: string }
  | { kind: "blob"; uri: string; mime_type?: string | null; blob_base64: string };

export interface McpInputRequired {
  prompt: string;
  inputSchema?: Record<string, unknown> | null;
}

export type McpResourceReadResponse =
  | { contents: McpResourceContent[] }
  | { input_required: McpInputRequired };

export interface McpPromptArgument {
  name: string;
  description?: string | null;
  required?: boolean;
}

export interface McpPromptDescriptor {
  name: string;
  title?: string | null;
  description?: string | null;
  arguments: McpPromptArgument[];
}

export type McpPromptMessageContent =
  | { type: "text"; text: string }
  | { type: "image"; data: string; mime_type: string }
  | { type: "audio"; data: string; mime_type: string }
  | { type: "resource_link"; uri: string; name: string }
  | { type: "embedded_resource"; resource: McpResourceContent };

export interface McpPromptMessage {
  role: "user" | "assistant";
  content: McpPromptMessageContent;
}

export interface McpPromptResult {
  description?: string | null;
  messages: McpPromptMessage[];
}

export type McpPromptGetResponse = McpPromptResult | { input_required: McpInputRequired };

// ── Labels / tones (pure) ──

export const MCP_RUNTIME_STATUS_LABELS: Record<McpRuntimeStatus, string> = {
  disconnected: "未连接",
  connecting: "连接中",
  ready: "就绪",
  degraded: "降级",
  unavailable: "不可用",
  misconfigured: "配置错误",
  disabled: "已禁用",
};

export type McpRuntimeTone = "default" | "success" | "warning" | "danger" | "info";

export const MCP_RUNTIME_STATUS_TONES: Record<McpRuntimeStatus, McpRuntimeTone> = {
  disconnected: "default",
  connecting: "info",
  ready: "success",
  degraded: "warning",
  unavailable: "danger",
  misconfigured: "danger",
  disabled: "default",
};

/**
 * Canonical transport label. Streamable HTTP (and its legacy `http`/`sse`
 * aliases) is a first-class, supported transport — no transport is described
 * as "not supported" anymore.
 */
export function mcpTransportRuntimeLabel(transport: string): string {
  switch (transport) {
    case "stdio":
      return "Stdio";
    case "streamable_http":
    case "http":
      return "Streamable HTTP";
    case "sse":
      return "SSE (legacy)";
    default:
      return transport || "未知";
  }
}

// ── Preview helpers (pure, display-bound, secret-safe) ──

export const EXTERNAL_PROMPT_WARNING = "来自外部 MCP Server 的内容，仅供预览。";

export const RESOURCE_TEXT_PREVIEW_MAX = 8000;

export type ResourcePreview =
  | { kind: "text"; text: string; truncated: boolean }
  | { kind: "blob"; mimeType: string | null; size: number };

/**
 * Reduce a resource content item to a safe, bounded preview.
 *
 * Text is length-capped for display. Blob content is NEVER surfaced: only the
 * MIME type and the encoded byte count are exposed — no base64, no decode, no
 * download, no execution.
 */
export function resourcePreview(content: McpResourceContent): ResourcePreview {
  if (content.kind === "blob") {
    return {
      kind: "blob",
      mimeType: content.mime_type ?? null,
      size: content.blob_base64.length,
    };
  }
  const truncated = content.text.length > RESOURCE_TEXT_PREVIEW_MAX;
  const text = truncated
    ? content.text.slice(0, RESOURCE_TEXT_PREVIEW_MAX) + "…"
    : content.text;
  return { kind: "text", text, truncated };
}

/** Render a prompt message as bounded plain text for preview. */
export function promptMessageText(message: McpPromptMessage): string {
  const content = message.content;
  switch (content.type) {
    case "text":
      return content.text;
    case "resource_link":
      return `[资源链接] ${content.uri}`;
    case "image":
      return "[图片] 未展开";
    case "audio":
      return "[音频] 未展开";
    case "embedded_resource": {
      const preview = resourcePreview(content.resource);
      return preview.kind === "text" ? preview.text : "[二进制资源] 未展开";
    }
  }
}

// ── API ──

export async function listMcpRuntimeServers(): Promise<McpServerRuntimeSummary[] | null> {
  const res = await request<{ servers: McpServerRuntimeSummary[] }>("GET", "/api/mcp/servers");
  return res?.servers ?? null;
}

export async function getMcpRuntimeServer(id: string) {
  return request<McpServerRuntimeSummary>("GET", `/api/mcp/servers/${id}`);
}

export async function refreshMcpRuntimeServer(id: string) {
  return request<McpServerRuntimeSummary>("POST", `/api/mcp/servers/${id}/refresh`);
}

export async function listMcpRuntimeTools(id: string): Promise<McpRuntimeTool[] | null> {
  const res = await request<{ tools: McpRuntimeTool[] }>("GET", `/api/mcp/servers/${id}/tools`);
  return res?.tools ?? null;
}

export async function listMcpResources(id: string): Promise<McpResourceDescriptor[] | null> {
  const res = await request<{ resources: McpResourceDescriptor[] }>(
    "GET",
    `/api/mcp/servers/${id}/resources`
  );
  return res?.resources ?? null;
}

export async function listMcpResourceTemplates(
  id: string
): Promise<McpResourceTemplate[] | null> {
  const res = await request<{ templates: McpResourceTemplate[] }>(
    "GET",
    `/api/mcp/servers/${id}/resource-templates`
  );
  return res?.templates ?? null;
}

export async function readMcpResource(id: string, uri: string) {
  return request<McpResourceReadResponse>("POST", `/api/mcp/servers/${id}/resources/read`, {
    uri,
  });
}

export async function listMcpPrompts(id: string): Promise<McpPromptDescriptor[] | null> {
  const res = await request<{ prompts: McpPromptDescriptor[] }>(
    "GET",
    `/api/mcp/servers/${id}/prompts`
  );
  return res?.prompts ?? null;
}

export async function getMcpPrompt(
  id: string,
  name: string,
  args: Record<string, unknown>
) {
  return request<McpPromptGetResponse>("POST", `/api/mcp/servers/${id}/prompts/get`, {
    name,
    arguments: args,
  });
}

File: backend/src/api/mcp_runtime.rs
// ============================================================
// MCP Runtime API — read-only runtime surface (control-session protected).
//
// No direct tool execution API is exposed here: MCP tools execute only through
// the Security Execution Gateway.
// ============================================================

use std::sync::Arc;

use axum::{
    extract::{Path, State},
    http::StatusCode,
    Json,
};
use serde::Deserialize;

use crate::mcp_runtime::{McpServerRuntime, McpTransportConfig};
use crate::server::AppServer;

fn server_dto(runtime: &McpServerRuntime) -> serde_json::Value {
    let transport = match &runtime.config {
        McpTransportConfig::Stdio { .. } => "stdio",
        McpTransportConfig::StreamableHttp { .. } => "http",
    };
    serde_json::json!({
        "id": runtime.server_id,
        "name": runtime.name,
        "transport": transport,
        "protocol_version": runtime.protocol_version,
        "status": runtime.status,
        "capabilities": runtime.capabilities,
        "tools_count": runtime.tools.len(),
        "resources_count": runtime.resources.len(),
        "resource_templates_count": runtime.resource_templates.len(),
        "prompts_count": runtime.prompts.len(),
        "last_refresh": runtime.last_refresh,
        "safe_error": runtime.last_error,
    })
}

pub async fn list_servers(State(server): State<Arc<AppServer>>) -> Json<serde_json::Value> {
    let runtimes = server.mcp_runtime_manager.list_servers();
    let views = runtimes.iter().map(|r| server_dto(r)).collect::<Vec<_>>();
    Json(serde_json::json!({ "servers": views }))
}

pub async fn get_server(
    State(server): State<Arc<AppServer>>,
    Path(id): Path<String>,
) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
    let runtime = server
        .mcp_runtime_manager
        .get_server(&id)
        .ok_or_else(|| (StatusCode::NOT_FOUND, "MCP server 不存在".to_string()))?;
    Ok(Json(server_dto(&runtime)))
}

pub async fn refresh_server(
    State(server): State<Arc<AppServer>>,
    Path(id): Path<String>,
) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
    server
        .mcp_runtime_manager
        .refresh_server(&id)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    let runtime = server.mcp_runtime_manager.get_server(&id).unwrap();
    Ok(Json(server_dto(&runtime)))
}

pub async fn list_tools(
    State(server): State<Arc<AppServer>>,
    Path(id): Path<String>,
) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
    let tools = server
        .mcp_runtime_manager
        .list_tools(&id)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    Ok(Json(serde_json::json!({ "tools": tools })))
}

pub async fn list_resources(
    State(server): State<Arc<AppServer>>,
    Path(id): Path<String>,
) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
    let resources = server
        .mcp_runtime_manager
        .list_resources(&id)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    Ok(Json(serde_json::json!({ "resources": resources })))
}

pub async fn list_resource_templates(
    State(server): State<Arc<AppServer>>,
    Path(id): Path<String>,
) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
    let templates = server
        .mcp_runtime_manager
        .list_resource_templates(&id)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    Ok(Json(serde_json::json!({ "templates": templates })))
}

#[derive(Deserialize)]
pub struct ReadResourceRequest {
    pub uri: String,
}

pub async fn read_resource(
    State(server): State<Arc<AppServer>>,
    Path(id): Path<String>,
    Json(body): Json<ReadResourceRequest>,
) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
    let cancel = tokio_util::sync::CancellationToken::new();
    match server
        .mcp_runtime_manager
        .read_resource(&id, &body.uri, &cancel)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
    {
        crate::mcp_runtime::McpOperationOutcome::Complete(contents) => {
            Ok(Json(serde_json::json!({ "contents": contents })))
        }
        crate::mcp_runtime::McpOperationOutcome::InputRequired(ir) => {
            Ok(Json(serde_json::json!({ "input_required": ir })))
        }
    }
}

pub async fn list_prompts(
    State(server): State<Arc<AppServer>>,
    Path(id): Path<String>,
) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
    let prompts = server
        .mcp_runtime_manager
        .list_prompts(&id)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
    Ok(Json(serde_json::json!({ "prompts": prompts })))
}

#[derive(Deserialize)]
pub struct GetPromptRequest {
    pub name: String,
    #[serde(default)]
    pub arguments: serde_json::Value,
}

pub async fn get_prompt(
    State(server): State<Arc<AppServer>>,
    Path(id): Path<String>,
    Json(body): Json<GetPromptRequest>,
) -> Result<Json<serde_json::Value>, (StatusCode, String)> {
    let cancel = tokio_util::sync::CancellationToken::new();
    match server
        .mcp_runtime_manager
        .get_prompt(&id, &body.name, body.arguments, &cancel)
        .await
        .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?
    {
        crate::mcp_runtime::McpOperationOutcome::Complete(result) => {
            Ok(Json(serde_json::json!(result)))
        }
        crate::mcp_runtime::McpOperationOutcome::InputRequired(ir) => {
            Ok(Json(serde_json::json!({ "input_required": ir })))
        }
    }
}
